test/functional/p2p_sendrecon.py

```diff
@@ -0,0 +1,139 @@
+    P2P_VERSION,
+)
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal
+
+class PeerNoVerack(P2PInterface):
+    def __init__(self, wtxidrelay=True):
+        super().__init__(wtxidrelay=wtxidrelay)
+
+    def on_version(self, message):
+        # Avoid sending verack in response to version.
+        # When calling add_p2p_connection, wait_for_verack=False must be set (see
+        # comment in add_p2p_connection).
+        if message.nVersion >= 70016 and self.wtxidrelay:
+            self.send_message(msg_wtxidrelay())
+
+class SendReconReceiver(P2PInterface):
+    def __init__(self):
+        super().__init__()
+        self.sendrecon_msg_received = None
+
+    def on_sendrecon(self, message):
+        self.sendrecon_msg_received = message
+
+def create_sendrecon_msg(initiator=True):
+    sendrecon_msg = msg_sendrecon()
+    sendrecon_msg.initiator = initiator
+    sendrecon_msg.responder = not initiator
+    sendrecon_msg.version = 1
+    sendrecon_msg.salt = 2
+    return sendrecon_msg
+
+class SendReconTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.extra_args = [['-txrecon']]
+
+    def run_test(self):
+        self.log.info('SENDRECON sent to an inbound')
+        peer = self.nodes[0].add_p2p_connection(SendReconReceiver(), send_version=True, wait_for_verack=True)
+        assert peer.sendrecon_msg_received
+        assert not peer.sendrecon_msg_received.initiator
+        assert peer.sendrecon_msg_received.responder
+        assert_equal(peer.sendrecon_msg_received.version, 1)
+
+        self.log.info('SENDRECON on pre-WTXID version should not be sent')
+        peer = self.nodes[0].add_p2p_connection(SendReconReceiver(), send_version=False, wait_for_verack=False)
+        pre_wtxid_version_msg = msg_version()
+        pre_wtxid_version_msg.nVersion = 70015
+        pre_wtxid_version_msg.strSubVer = P2P_SUBVERSION
+        pre_wtxid_version_msg.nServices = P2P_SERVICES
+        pre_wtxid_version_msg.relay = 1
+        peer.send_message(pre_wtxid_version_msg)
+        peer.wait_for_verack()
+        assert not peer.sendrecon_msg_received
+
+        self.log.info('SENDRECON for fRelay=false should not be sent')
+        peer = self.nodes[0].add_p2p_connection(SendReconReceiver(), send_version=False, wait_for_verack=False)
+        no_txrelay_version_msg = msg_version()
+        no_txrelay_version_msg.nVersion = P2P_VERSION
+        no_txrelay_version_msg.strSubVer = P2P_SUBVERSION
+        no_txrelay_version_msg.nServices = P2P_SERVICES
+        no_txrelay_version_msg.relay = 0
+        peer.send_message(no_txrelay_version_msg)
+        peer.wait_for_verack()
+        assert not peer.sendrecon_msg_received
+
+        self.log.info('valid SENDRECON received')
+        peer = self.nodes[0].add_p2p_connection(PeerNoVerack(), send_version=True, wait_for_verack=False)
+        peer.send_message(create_sendrecon_msg())
```

Comment:
6ec8628181bb9342c338d180acc96f5a48185527
It may be helpful to sync here to make sure the disconnect isn't caused by the first sendrecon

`self.wait_until(lambda : self.nodes[0].getpeerinfo()[-1]["bytesrecv_per_msg"]["sendrecon"])`